nixos/modules/services/misc/workout-tracker.nix

```diff
@@ -0,0 +1,123 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+
+let
+  inherit (lib) types;
+  cfg = config.services.workout-tracker;
+in
+
+{
+  options = {
+    services.workout-tracker = {
+      enable = lib.mkEnableOption "A workout tracking web application for personal use (or family, friends), geared towards running and other GPX-based activities";
+
+      package = lib.mkPackageOption pkgs "workout-tracker" { };
+
+      user = lib.mkOption {
+        type = types.str;
+        default = "workout-tracker";
+        description = "User account under which Workout-tracker runs.";
+      };
+
+      group = lib.mkOption {
+        type = types.str;
+        default = "workout-tracker";
+        description = "Group under which Workout-tracker runs.";
+      };
+
+      environmentFile = lib.mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        example = "/run/keys/workout-tracker-env";
+        description = ''
+          An environment file as defined in {manpage}`systemd.exec(5)`.
+
+          Secrets like `WT_JWT_ENCRYPTION_KEY` may be passed to the service without adding them
+          to the world-readable Nix store.
+        '';
+      };
+
+      settings = lib.mkOption {
+        type = lib.types.submodule {
+          freeformType =
+            with lib.types;
+            attrsOf (
+              let
+                typeList = [
+                  bool
+                  float
+                  int
+                  str
+                  path
+                  package
+                ];
+              in
+              oneOf (
+                typeList
+                ++ [
+                  (listOf (oneOf typeList))
+                  (attrsOf (oneOf typeList))
+                ]
+              )
+            );
+
+          options = {
+            address = lib.mkOption {
+              type = types.str;
+              default = "localhost";
+              description = "Web interface address.";
+            };
+
+            port = lib.mkOption {
+              type = types.port;
+              default = 8080;
+              description = "Web interface port.";
+            };
+
+            stateDir = lib.mkOption {
+              default = "/var/lib/workout-tracker";
+              type = types.str;
+              description = "Workout-tracker data directory.";
+            };
+          };
```

Comment:
I think I gave you a a bit wrong suggestion.
The `stateDir` shouldn't be part of `services.workout-tracker.settings`. 
For `address` and `port`, there are two possible ways to handle that
1. Replace them with one `WT_BIND` option that has `localhost:8000` as a default
2. Re-move them out of `services.workout-tracker.settings`. Then instead of setting `WT_BIND` in line 107, set `cfg.settings.WT_BIND = "${cfg.settings.address}:${toString cfg.settings.port}"` somewhere in the `config` block.

I see advantages and disadvantages with both options. Feel free to decide which you think is better (I would probably choose 2., but very light preference)

Reply:
Went with option 2, reworked things fair bit, used the `default` field to set it, but user can override. WDYT?

Reply:
This looks good to me! I would probably remove `WT_BIND` from the example, as a user can think that they should set it

Reply:
This service is intended to be served from an always on device, something that is more likely than not to be user's non-primary, non-interactive device. I think it will help to show people via example that this is what they can do to make it available over network. Also, `options.nixos.org` renders both default and example, so it should be fine.

I will remove it if you think that is a good idea, but this is my rationale.

Reply:
But shouldn't they then change `services.workout-tracker.address` instead of `services.workout-tracker.settings.WT_BIND`

Reply:
You need also a `defaultText` for `.settings` so that the NixOS docs build

Reply:
fair point, removed.

Also having it in `default` causes issues with manual generation. So now setting it via service env, which from what I understand takes precedence over what gets added to EnvironmentFile.

Reply:
> You need also a `defaultText` for `.settings` so that the NixOS docs build

Its fixed, just needed to not refer to `cfg` in default text.